docs: Clarify important points about parametric types

adigitoleo opened the pull request:
Closes #43811.

- The intro section felt a bit long-winded, I've made some changes there first.
- Clarify that `typeof` returns the concrete type
- Rename Type Declarations section to Type Annotations,
  avoid confusion with Declaring Types section and distinguish use of "type declaration"
  to mean "declaring new types"
- Removed some of the jargon and wikipedia links to make room for
  a brief alternative `Point{T1,T2}` demonstration.
- Shifted some paragraphs around to reflect their importance,
  and changed the wording in some places.
- Rename type declaration -> annotation in other places (docstrings/comments)

adigitoleo committed:
docs: Clarify important points about parametric types

Closes #43811.

- The intro section felt a bit long-winded, I've made some changes there first.
- Clarify that `typeof` returns the concrete type
- Rename Type Declarations section to Type Annotations,
  avoid confusion with Declaring Types section and distinguish use of "type declaration"
  to mean "declaring new types"
- Removed some of the jargon and wikipedia links to make room for
  a brief alternative `Point{T1,T2}` demonstration.
- Shifted some paragraphs around to reflect their importance,
  and changed the wording in some places.
- Rename type declaration -> annotation in other places (docstrings/comments)

adigitoleo commented:
Couldn't find a good place to add a complete custom type family example, and the section is already quite long. Maybe it can go somewhere else eventually.

adigitoleo commented:
One other thing I left out compared to the  linked issue was talking about DataType, because that has it's own section in Declaring Types.

I do think that Declaring Types needs a bit of work as well, but I'll leave it for another PR.

adigitoleo commented:
Happy to revert the formatting if it helps. I automatically adopted my markdown habits of breaking at sentence clauses, which I find to produce comfy diffs (except for the first time it's done, like here).

Here's the built html of that page, if it helps: https://x0.at/Bw6y.txt

StefanKarpinski commented:
I also prefer that style, but it makes the diff here too big and hard to read. You could separate the change into purely formatting and actual content changes as well and then I can just review the content change.

adigitoleo committed:
docs: Tame formatting changes

adigitoleo commented:
Whoops, shouldn't have force-pushed?

fingolfin commented:
Overall this looks quite good to me! It is a pity that it sat around here for almost 2 years :-(. @adigitoleo sorry for that, if you are still willing to work on it, perhaps you can resolve the conflict?

@StefanKarpinski I think this is relatively nice to review now, perhaps you'd be willing to have another look at it?

stevengj commented on doc/src/manual/types.md:
```diff
@@ -1,73 +1,59 @@
+are the types of any values ever restricted.
+
+Julia's type system is dynamic, but gains some of the advantages of static type systems
+by allowing optional type annotations. This can be of great assistance
```
Maybe:

> Type annotations are important for defining efficient new data structures, can help enforce correctness and clarity, and can sometimes improve efficiency (although in many contexts the Julia compiler infers types automatically). However, type annotations also play a central role in Julia because they control method dispatch: functions can be extended to new behaviors for different argument types.

stevengj commented on doc/src/manual/types.md:
```diff
@@ -1,73 +1,59 @@
+The default behavior in Julia when type annotations are omitted is to allow values to be of any type.
+When additional expressiveness is needed, however,
+it is easy to gradually introduce explicit annotations into previously "untyped" code.
+Adding annotations serves three primary purposes: to take advantage
+of Julia's powerful multiple-dispatch mechanism, to improve human readability,
+and to catch programmer errors.
```
```suggestion
Adding annotations serves four primary purpose: to take advantage
of Julia's powerful multiple-dispatch mechanism, to improve human readability,
to catch programmer errors, and to assist the compiler (especially when defining new types).
```
[Nobody expects the Spanish Inquisition!](https://www.youtube.com/watch?v=Cj8n4MfhjUc)

stevengj commented on doc/src/manual/types.md:
```diff
@@ -78,7 +64,10 @@ julia> (1+2)::Int
+When the type on the right is concrete,
+the value on the left must have that type as its implementation.
+When the type is abstract,
+it suffices for the value to be implemented by a concrete type that is a subtype of the abstract type.
```
```suggestion
it suffices for the value to have a concrete type that is a subtype of the abstract type.
```

stevengj commented on doc/src/manual/types.md:
```diff
@@ -595,13 +592,13 @@ julia> struct Point{T}
-is now a usable concrete type:
+This declaration defines a new parametric type, `Point{T}`,
+holding two "coordinates" of the same type `T`.
+The parameter `T` is a placeholder, which will be replaced by a particular concrete type
```
```suggestion
The parameter `T` is a placeholder, which will be replaced by a particular type
```
It doesn't have to be concrete, e.g. you can have `Point{Real}`.

adigitoleo committed:
docs: Apply review suggestions

adigitoleo commented:
Thanks for the review. Conflicts should be resolved now. I've addressed the comments, but I'm not sure of the etiquette: should I mark conversations as resolved or leave that to maintainers?

oscardssmith commented:
I think this is ready to merge, right?

adigitoleo commented:
Just rebased on lates master, haven't looked at the bulk of the changes in a while but it has been reviewed before. Should be good to go from my side.

stevengj commented on doc/src/manual/types.md:
```diff
@@ -1,66 +1,59 @@
+and dynamic type systems, where types are only computed at run time,
+when the actual values manipulated by the program are available.
+Statically typed languages typically offer faster execution,
+at the cost of type annotations which must be explicitly added by the programmer.
```
Explicit annotations are not required in many statically typed languages, e.g. with [Hindley–Milner type systems](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system), which (like Julia) use type inference.

The difference, as @StefanKarpinski has often put it, is that in a static language it is an error if type inference fails, whereas in a dynamic language like Julia the inference is just an optimization.

adigitoleo replied:
Good points, thanks. How about this for the first paragraph:

> Types in Julia establish distinctions between different kinds of data, and are used by the compiler to infer the intended use of those data. Programming languages have traditionally employed one of two quite different type systems: static type systems, where expressions must have a computable type before the execution of the program, and dynamic type systems, where types are only computed at run time, when the actual values manipulated by the program are available. Statically typed languages typically offer faster execution of programs, while treating any data of unknown or invalid type as an error. In these languages, the programmer is responsible for ensuring the validity of all types of data that the program may encounter. Dynamically typed languages, on the other hand, do not explicitly check types, and types of data only become invalid when they fail to support run-time operations.

stevengj replied:
> Statically typed languages typically offer faster execution of programs

They are not any faster than type-stable code in Julia.   The main argument for static typing, IMO, is safety — by being stricter about types, the compiler can catch more bugs.

But the pros and cons of static vs. dynamic languages is a longstanding debate and a bit of a can of worms here.

adigitoleo replied:
I guess by "typically" I probably meant "usually", but I can understand if this unintentionally suggests that Julia is not performant. I'll think about a better wording here.

fingolfin replied:
I don't think "usually" is any better

adigitoleo replied:
> Statically typed languages offer stricter guidance for avoiding unsafe code, such as treating any data of unknown or invalid type as an error. In these languages, the programmer is responsible for ensuring the validity of all types of data that the program may encounter. Writing programs in statically typed languages is, therefore, more laborious, and interactive exploration of data at run-time becomes more difficult. Dynamically typed languages, on the other hand, do not explicitly check types, and types of data only become invalid when they fail to support run-time operations. This provides a more flexible and productive programming experience, at the cost of a greater potential for programmer errors to cause bugs during execution.

?

stevengj commented on doc/src/manual/types.md:
```diff
@@ -1,66 +1,59 @@
+Statically typed languages typically offer faster execution,
+at the cost of type annotations which must be explicitly added by the programmer.
+Dynamic typing, on the other hand, allows for polymorphism,
+which is the ability to write code that can operate on different types.
```
Statically typed languages also have polymorphism.

adigitoleo replied:
See above suggestion, wherein I removed the misleading reference to polymorphism.

stevengj commented on doc/src/manual/types.md:
```diff
@@ -567,26 +563,34 @@ See [this FAQ entry](@ref faq-nothing) for more information.
-Type parameters are introduced immediately after the type name, surrounded by curly braces:
+!!! note
+    For any parametric composite type,
+    both [`isabstracttype`](@ref) and [`isconcretetype`](@ref) will return false.
```
```suggestion
    both [`isabstracttype`](@ref) and [`isconcretetype`](@ref) will return `false`.
```

adigitoleo committed:
fix: Remove unintended whitespace change

adigitoleo committed:
Merge branch 'master' into docs-man-types

adigitoleo committed:
docs: Remove misleading reference to polymorphism

adigitoleo committed:
fix: Add missing backticks

adigitoleo commented:
I have updated the branch with commits to address the reviews by @stevengj, and the full html folder of the built docs should be downloadable from [this link](https://cryptpad.fr/file/#/2/file/rCHwHbJoYyqxPkTgn-X0x17o/), where the `manual/types.html` file contains the most significant changes.

akdienes commented on doc/src/manual/types.md:
```diff
@@ -78,7 +70,10 @@ julia> (1+2)::Int
 This allows a type assertion to be attached to any expression in-place.
-
+When the type on the right is concrete,
+the value on the left must have that type as its implementation.
```
```suggestion
```

this is true for dispatch purposes, but not true in general in cases like `x::Int = 'a'`

adienes commented on doc/src/manual/types.md:
```diff
@@ -78,7 +70,10 @@ julia> (1+2)::Int
 This allows a type assertion to be attached to any expression in-place.
-
+When the type on the right is concrete,
+the value on the left must have that type as its implementation.
```
```suggestion
```

this is true for dispatch purposes, but not true in general in cases like `x::Int = 'a'`

adigitoleo replied:
```
When the type on the right is concrete,
it must be [`convert`](@ref)able to the type implemented by the value on the left.
```
?

adigitoleo replied:
This is covered by the rest of the paragraph:

> When appended to a variable on the left-hand side of an assignment, or as part of a `local` declaration, the `::` operator means something a bit different: it declares the variable to always have the specified type, like a type declaration in a statically-typed language such as C. Every value assigned to the variable will be converted to the declared type using [`convert`](@ref):

adienes replied:
woops, yes

adienes commented on doc/src/manual/types.md:
```diff
@@ -1,66 +1,58 @@
+
+All values in a Julia program belong to exactly one concrete type.
+All types belong to a single type tree, i.e. they are all at least related to the [`Any`](@ref) type.
+User-defined types are treated the same as built-in types.
```
```suggestion
```

either this is wrong, or it needs more explanation about what "the same" means here. because there are certainly built-in types which get very special treatment

adienes commented on doc/src/manual/types.md:
```diff
@@ -567,26 +562,34 @@ See [this FAQ entry](@ref faq-nothing) for more information.
+whereas [parametric abstract types](#Parametric-Abstract-Types)
+declare a family of [abstract types](#man-abstract-types).
+Other kinds of built-in parametric types such as tuple types are discussed in this section as well.
+Parametric types can also be used in [type annotations](#Type-Annotations).
```
```suggestion
Parametric types can also be used in [type annotations](#Type-Annotations) when the parameters are specified.
```
`x::Foo{T} = 1` is not possible.

adienes commented on doc/src/manual/types.md:
```diff
@@ -567,26 +562,34 @@ See [this FAQ entry](@ref faq-nothing) for more information.
+
+!!! warning
+    Type parameters are invariant: even though `Float64 <: Real`,
+    the relation `Foo{Float64} <: Foo{Real}` does NOT hold for any parametric type `Foo`.
```
```suggestion
    the relation `Foo{Float64} <: Foo{Real}` does NOT hold for any parametric type `Foo`.
```

of course except for `Tuple`

adigitoleo replied:
I didn't know about that, thanks. Is there a reason why that is the case that we can/should highlight here?

adienes replied:
I'm not familiar with all the details but my understanding is that the reason for covariance is largely vestigial; it made dispatch easier in the early days of Julia or something


Tuple type covariance is described here in the docs https://docs.julialang.org/en/v1/manual/types/#Tuple-Types-1

and there is an open issue discussing the potential to switch to invariance in a hypothetical breaking release of Julia (although as a gambling man, I wouldn't put my money on that change ever happening) https://github.com/JuliaLang/julia/issues/24614

I would think a quick link to the docs is probably sufficient

adienes commented on doc/src/manual/functions.md:
```diff
@@ -106,9 +106,9 @@ the call site that at least one of the arguments (often the first one) is being
+## Argument-type annotations
 
-You can declare the types of function arguments by appending `::TypeName` to the argument name, as usual for [Type Declarations](@ref) in Julia.
+You can declare the types of function arguments by appending `::TypeName` to the argument name, as usual for [Type Annotations](@ref) in Julia.
```
```suggestion
You can annotate the types of function arguments by appending `::TypeName` to the argument name, as usual for [Type Annotations](@ref) in Julia.
```

if the noun is changing `declaration` --> `annotation` then the verb should change `declare` --> `annotate`

adienes commented on doc/src/manual/performance-tips.md:
```diff
@@ -250,6 +250,13 @@ better than `IdDict{Type, Vector}`
+
 ### Avoid fields with abstract type
 
 Types can be declared without specifying the types of their fields:
```
```suggestion
Types can be annotated without specifying the types of their fields:
```

adienes commented on doc/src/manual/types.md:
```diff
@@ -1,66 +1,58 @@
 
 Julia's type system is designed to be powerful and expressive, yet clear, intuitive and unobtrusive.
 Many Julia programmers may never feel the need to write code that explicitly uses types. Some
 kinds of programming, however, become clearer, simpler, faster and more robust with declared types.
```
```suggestion
kinds of programming, however, become clearer, simpler, faster and more robust with annotated types.
```

adienes commented on doc/src/manual/types.md:
```diff
@@ -78,7 +70,10 @@ julia> (1+2)::Int
+When the type is abstract,
+it suffices for the value to have a concrete type that is a subtype of the abstract type.
 When appended to a variable on the left-hand side of an assignment, or as part of a `local` declaration,
 the `::` operator means something a bit different: it declares the variable to always have the
```
```suggestion
the `::` operator means something a bit different: it annotates the variable to always have the
```

adienes commented on doc/src/manual/types.md:
```diff
@@ -595,13 +598,13 @@ julia> struct Point{T}
+holding two "coordinates" of the same type `T`.
+The parameter `T` is a placeholder, which will be replaced by a particular type
+when an instance of this type is created.
+Thus, this single declaration actually declares an unlimited number of types:
```
```suggestion
Thus, this single annotation actually declares an unlimited number of types:
```

adienes commented on doc/src/manual/types.md:
```diff
@@ -1,66 +1,58 @@
+User-defined types are treated the same as built-in types.
+For additional flexibility, types can be defined as concrete, abstract or parametric.
+Their relation to other types must be explicitly declared.
+One particularly distinctive feature of Julia's type system is that only abstract types can be used as supertypes for a family of related types.
```
```
julia> struct Foo{T} end

julia> Foo{Int} <: Foo
true

julia> isabstracttype(Foo)
false
```

adigitoleo committed:
Apply suggestions from code review

Co-authored-by: adienes <51664769+adienes@users.noreply.github.com>

adigitoleo committed:
Apply suggestions from code review

Co-authored-by: adienes <51664769+adienes@users.noreply.github.com>

adigitoleo commented:
Hi. Is there still interest in getting this merged? I should be able to take a look in March if more changes are required.

adienes commented:
personally I think it looks good to go. I'll mark `triage` to see if the next call can take a quick look, not because I think there's anything controversial here but only because rewriting docs on such a fundamental language feature is by nature going to be a sensitive operation.

LilithHafner committed:
Merge branch 'master' into docs-man-types

LilithHafner commented on doc/src/manual/performance-tips.md:
```diff
@@ -257,9 +257,16 @@ better than `IdDict{Type, Vector}`
 ### Avoid fields with abstract type
 
-Types can be declared without specifying the types of their fields:
+Types can be annotated without specifying the types of their fields:
```
Declared is correct here.

LilithHafner commented on doc/src/manual/performance-tips.md:
```diff
@@ -257,9 +257,16 @@ better than `IdDict{Type, Vector}`
 
 See also the discussion under [Parametric Types](@ref).
 
+## Type annotations
+
+In many languages with optional type annotations, adding such annotations is the principal way to
+make code run faster. This is *not* the case in Julia. In Julia, the compiler generally knows
+the types of all function arguments, local variables, and expressions. However, there are a few
+specific instances where annotations are helpful.
+
 ### Avoid fields with abstract type
```
This is already present here:
https://github.com/JuliaLang/julia/blob/4a6ada6eb36eaeb9bde9b94c01bcf6f0e3ce7ae6/doc/src/manual/performance-tips.md?plain=1#L209-L214

adigitoleo committed:
Address code review

adigitoleo commented:
Thanks @LilithHafner, good catch on those and sorry that it took a little while for me to circle back to this.

LilithHafner commented:
Thanks, we'll discuss this at Triage per @adienes's request!

DilumAluthge commented:
@LilithHafner Did this ever get discussed at triage?

I see the triage label here, but I don't see a comment summarizing triage's feedback.

LilithHafner commented:
@DilumAluthge It did not, to my knowledge.
